include/seahorn/SolverBmc.hh

```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+#include "llvm/IR/Function.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "boost/logic/tribool.hpp"
+
+#include "seahorn/Expr/Expr.hh"
+#include "seahorn/Expr/Smt/EZ3.hh"
+#include "seahorn/Expr/Smt/Solver.hh"
+#include "seahorn/Expr/Smt/Z3SolverImpl.hh"
+
+#include "seahorn/Analysis/CutPointGraph.hh"
+#include "seahorn/OperationalSemantics.hh"
+
+namespace seahorn {
+using namespace expr;
+
+namespace solver_bmc_impl {
+/// true if I is a call to a void function
+bool isCallToVoidFn(const llvm::Instruction &I);
+/// computes an implicant of f (interpreted as a conjunction) that
+/// contains the given model
+void get_model_implicant(const ExprVector &f, solver::Solver::model_ref model,
+                         ExprVector &out, ExprMap &active_bool_map);
+// out is a minimal unsat core f based on assumptions
+void unsat_core(solver::Solver &solvr, const ExprVector &f, bool simplify,
+                ExprVector &out);
+} // namespace solver_bmc_impl
+
+class SolverBmcTrace;
+class SolverBmcEngine {
+protected:
+  /// symbolic operational semantics
+  OperationalSemantics &m_sem;
+  /// \brief Context for OperationalSemantics
+  OpSemContextPtr m_semCtx;
+
+  /// expression factory
+  ExprFactory &m_efac;
+
+  /// last result
+  solver::SolverResult m_result;
+
+  /// cut-point trace
+  SmallVector<const CutPoint *, 8> m_cps;
+
+  /// symbolic states corresponding to m_cps
+  std::vector<SymStore> m_states;
+  /// edge-trace corresponding to m_cps
+  SmallVector<const CpEdge *, 8> m_edges;
+
+  const CutPointGraph *m_cpg;
+  const llvm::Function *m_fn;
+
+  // main solver
+  solver::SolverKind m_solver_kind;
+  std::unique_ptr<solver::Solver> m_new_smt_solver;
+  // simplifying solver that only relies on z3
+  std::unique_ptr<solver::Solver> m_simp_smt_solver;
+
+  SymStore m_ctxState;
+  /// path-condition for m_cps
+  ExprVector m_side;
+
+public:
+  SolverBmcEngine(OperationalSemantics &sem, EZ3 &zctx,
+                  solver::SolverKind solver_kind = solver::SolverKind::Z3);
+  void addCutPoint(const CutPoint &cp);
+
+  virtual OperationalSemantics &sem() { return m_sem; }
+
+  EZ3 &zctx() {
+    if (m_solver_kind == solver::SolverKind::Z3) {
+      auto &solver = static_cast<solver::z3_solver_impl &>(*m_new_smt_solver);
+      return solver.get_context();
+    } else if (m_new_smt_solver->get_kind() == solver::SolverKind::YICES2) {
+      auto m_zctx = new EZ3(m_efac);
```

Comment:
this is a memory leak. someone must own this allocation. currently it is never deallocated